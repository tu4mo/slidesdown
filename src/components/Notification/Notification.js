import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'

const Notification = ({ children, className, timeout }) => {
  const [hasTimedOut, setHasTimedOut] = useState(false)

  useEffect(() => {
    const timer = setTimeout(() => setHasTimedOut(true), timeout || 15000)
    return () => clearTimeout(timer)
  }, [])

  return !hasTimedOut && <div className={className}>{children}</div>
}

Notification.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  timeout: PropTypes.number
}

const StyledNotification = styled(Notification)`
  color: ${props => props.theme.colors.purple};

  ${props =>
    props.slideDown &&
    css`
      animation: slide-down 1s ease-out;

      @keyframes slide-down {
        0% {
          opacity: 0;
          transform: translateY(-1rem);
        }
        100% {
          opacity: 1;
          transform: translateY(0);
        }
      }
    `};
`

StyledNotification.propTypes = {
  slideDown: PropTypes.bool
}

export default StyledNotification
