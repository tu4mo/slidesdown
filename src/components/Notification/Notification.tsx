import { ReactNode, useEffect, useState } from 'react'
import styled, { css, keyframes } from 'styled-components'

type Props = {
  children: ReactNode
  className?: string
  timeout?: number
}

const Notification = ({ children, className, timeout = 15000 }: Props) => {
  const [hasTimedOut, setHasTimedOut] = useState(false)

  useEffect(() => {
    const timer = setTimeout(() => setHasTimedOut(true), timeout)
    return () => clearTimeout(timer)
  }, [timeout])

  return !hasTimedOut ? <div className={className}>{children}</div> : null
}

const slideDown = keyframes`
  0% {
    opacity: 0;
    transform: translateY(-1rem);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
`

const StyledNotification = styled(Notification)<{ $slideDown?: boolean }>`
  ${(props) =>
    props.$slideDown &&
    css`
      animation: ${slideDown} 1s ease-out;
    `};

  color: var(--color-purple);
`

export { StyledNotification as Notification }
