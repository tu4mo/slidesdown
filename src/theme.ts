export default {
  breakpoints: {
    md: '576px',
    xl: '768px'
  },
  colors: {
    background: '#edf1f5',
    border: '#e4e4e8',
    dimText: '#a4a4a8',
    lightText: '#646468',
    magenta: '#c86dd7',
    purple: '#3023ae'
  },
  fonts: {
    default: 'Rubik',
    monospace: 'Inconsolata'
  }
}
