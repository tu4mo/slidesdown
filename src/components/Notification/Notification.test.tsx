import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { ThemeProvider } from 'styled-components'

import theme from '../../theme'
import Notification from './Notification'

jest.useFakeTimers()

const component = renderer.create(
  <ThemeProvider theme={theme}>
    <Notification>Test</Notification>
  </ThemeProvider>
)

it('renders correctly', () => {
  expect(component.toJSON()).toMatchSnapshot()
})

// Disabled until there's a way to test useEffect
// it('renders nothing after time has passed', () => {
//   jest.runAllTimers()
//   expect(setTimeout).toHaveBeenCalledTimes(1)
//   expect(component.toJSON()).toMatchSnapshot()
// })
