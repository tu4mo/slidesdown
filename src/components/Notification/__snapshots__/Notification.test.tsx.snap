// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  color: #3023ae;
}

<div
  className="c0"
>
  Test
</div>
`;
